infrastructure/filefilter/filter.go

```diff
@@ -0,0 +1,215 @@
+package filefilter
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/rs/zerolog/log"
+	ignore "github.com/sabhiram/go-gitignore"
+	"gopkg.in/yaml.v3"
+)
+
+func FindNonIgnoredFiles(rootFolder string) <-chan string {
+	return newRepoIgnoresFilter(rootFolder).
+		findNonIgnoredFiles()
+}
+
+type repoIgnoresFilter struct {
+	// The path to the root of the repository
+	repoRoot               string
+	ignoreFiles            []string
+	globsPerFolder         map[string][]string
+	ignoreCheckerPerFolder map[string]ignore.IgnoreParser
+}
+
+func newRepoIgnoresFilter(rootFolder string) *repoIgnoresFilter {
+	return &repoIgnoresFilter{
+		repoRoot:               rootFolder,
+		ignoreFiles:            []string{".gitignore", ".dcignore", ".snyk"},
+		globsPerFolder:         make(map[string][]string),
+		ignoreCheckerPerFolder: make(map[string]ignore.IgnoreParser),
+	}
+}
+
+func (f *repoIgnoresFilter) findNonIgnoredFiles() <-chan string {
+	resultsCh := make(chan string)
+	var wg sync.WaitGroup
+	go func() {
+		defer close(resultsCh)
+		// When walking a directory, the iteration is the directory itself (`path` parameter is the directory path).
+		// The ignore files are immediately parsed, and the iterations of the files come after the ignore rules have been loaded
+		err := filepath.WalkDir(f.repoRoot, func(path string, dirEntry os.DirEntry, err error) error {
+			if err != nil {
+				return filepath.SkipAll
+			}
+			if dirEntry == nil {
+				return nil
+			}
+
+			if dirEntry.IsDir() {
+				// The following optimization has been removed because it doesn't take into account negation rules:
+				// For example:
+				// .ignoreme
+				// !.ignoreme/keepme.txt
+				//
+				// This implementation would skip the entire folder and not return keepme.txt:
+				//
+				// ```
+				//parentFolderPath := filepath.Dir(path)
+				//if path != rootFolder && ignoreCheckerPerFolder[parentFolderPath].MatchesPath(path) {
+				//	return filepath.SkipDir
+				//}
+				// ```
+				globs := f.collectGlobs(path)
+				f.globsPerFolder[path] = globs
+				f.ignoreCheckerPerFolder[path] = ignore.CompileIgnoreLines(globs...)
+				return nil
+			} else {
+				folderPath := filepath.Dir(path)
+				checker := f.ignoreCheckerPerFolder[folderPath]
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					if !checker.MatchesPath(path) {
+						resultsCh <- path
+					}
+				}()
+			}
+
+			return nil
+		})
+
+		wg.Wait()
+
+		if err != nil {
+			log.Err(err).Msg("Error during filepath.WalkDir")
+		}
+	}()
+
+	return resultsCh
+}
+
+func (f *repoIgnoresFilter) collectGlobs(path string) []string {
+	var globs []string
+	if path != f.repoRoot {
+		globs = append(globs, f.globsPerFolder[filepath.Dir(path)]...)
+	} else {
+		defaultGlobs := []string{"**/.git/**", "**/.svn/**", "**/.hg/**", "**/.bzr/**", "**/.DS_Store/**"}
+		globs = append(globs, defaultGlobs...)
+	}
+
+	for _, ignoreFile := range f.ignoreFiles {
+		ignoreFilePath := filepath.Join(path, ignoreFile)
```

Comment:
where do we report the progress of the filewalking?

Reply:
We can't report the progress percentage now. Previously we had one walk where we collect all the ignore rules and count the files, and then a 2nd walk where we filter the files. The progress of the 2nd walk was reported and the progress percentage was the count of the walked files divided by the total file count. 

Now we only have one walk, so we can't really know how many files are left.

We still have other parts where progress is reported, for example when uploading/analyzing/etc

Reply:
Then we need at least a progress message that is unquantifiable that says "Collecting files"

Reply:
Done